pkg/util/deployment/deployment.go

```diff
@@ -142,6 +144,136 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	for _, rs := range rsList {
+		// Add pod-template-hash information if it's not in the rs
+		if !labelsutil.SelectorHasLabel(rs.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey) {
+			updatedRS, err := addHashKeyToReplicaSet(deployment, c, rs, getPodList)
+			if err != nil {
+				return nil, err
+			}
+			syncedRSList = append(syncedRSList, *updatedRS)
+		}
+		syncedRSList = append(syncedRSList, rs)
+	}
+	return syncedRSList, nil
+}
+
+// addHashKeyToReplicaSet adds pod-template-hash information to the given rs with the following steps:
+// 1. Add hash label to the rs's pod template
+// 2. Add hash label to all pods this rs owns
+// 3. Add hash label to the rs's selector
+// 4. Clean up all pods this rs owns but without the hash label (orphaned pods)
+func addHashKeyToReplicaSet(deployment extensions.Deployment, c clientset.Interface, rs extensions.ReplicaSet, getPodList podListFunc) (*extensions.ReplicaSet, error) {
+	if labelsutil.SelectorHasLabel(rs.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey) {
+		return &rs, nil
+	}
+	namespace := deployment.Namespace
+	hash := podutil.GetPodTemplateSpecHash(*rs.Spec.Template)
+	// 1. Add hash template label to the rs. This ensures that any newly created pods will have the new label.
+	updatedRS, err := updateRSWithRetries(c.Extensions().ReplicaSets(namespace), &rs, func(updated *extensions.ReplicaSet) {
+		updated.Spec.Template.Labels = labelsutil.AddLabel(updated.Spec.Template.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. Update all pods managed by the rs to have the new hash label, so they are correctly adopted.
+	selector, err := unversioned.LabelSelectorAsSelector(updatedRS.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	podList, err := getPodList(namespace, options)
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		delay, maxRetries := 3, 3
+		for i := 0; i < maxRetries; i++ {
+			_, err = c.Core().Pods(namespace).Update(&pod)
+			if err != nil {
+				time.Sleep(time.Second * time.Duration(delay))
+				delay *= delay
+			} else {
+				break
+			}
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// 3. Update rs selector
+	// Copy the old selector, so that we can scrub out any orphaned pods
+	oldSelector := updatedRS.Spec.Selector
+	// Update the selector of the rs so it manages all the pods we updated above
+	if updatedRS, err = updateRSWithRetries(c.Extensions().ReplicaSets(namespace), updatedRS, func(updated *extensions.ReplicaSet) {
+		updated.Spec.Selector = labelsutil.AddLabelToSelector(updated.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey, hash)
+	}); err != nil {
+		return nil, err
+	}
+
+	// 4. Clean up any orphaned pods that don't have the new label, this can happen if the rs manager
```

Comment:
Since we already account for orphaned pods, we could move pod relabeling at 1. and collapse updating the rs podtemplate and selector in one update?